pkg/reconciler/ingress/config/gateway.go

```diff
@@ -0,0 +1,148 @@
+
+	visibilityConfigKey = "visibility"
+
+	// defaultGatewayClass is the gatewayclass name for the gateway.
+	defaultGatewayClass = "istio"
+
+	// defaultIstioGateway is the default gateway.
+	defaultIstioGateway = "istio-system/test-gateway"
+
+	// defaultIstioLocalGateway is the default local gateway:
+	defaultIstioLocalGateway = "istio-system/test-local-gateway"
+
+	// defaultLocalGatewayService holds the default local gateway service.
+	defaultLocalGatewayService = "istio-system/knative-local-gateway"
+
+	// defaultGatewayService is the default gateway service.
+	defaultGatewayService = "istio-system/istio-ingressgateway"
+)
+
+type GatewayConfig struct {
+	GatewayClass string `json:"class,omitempty"`
+	Gateway      string `json:"gateway,omitempty"`
+	Service      string `json:"service,omitempty"`
+}
+
+// Gateway maps gateways to routes by matching the gateway's
+// label selectors to the route's labels.
+type Gateway struct {
+	// Gateways map from gateway to label selector.  If a route has
+	// labels matching a particular selector, it will use the
+	// corresponding gateway.  If multiple selectors match, we choose
+	// the most specific selector.
+	Gateways map[v1alpha1.IngressVisibility]*GatewayConfig
+}
+
+// NewGatewayFromConfigMap creates a Gateway from the supplied ConfigMap
+func NewGatewayFromConfigMap(configMap *corev1.ConfigMap) (*Gateway, error) {
+	v, ok := configMap.Data[visibilityConfigKey]
+	if !ok {
+		// These are the defaults.
+		return &Gateway{
+			Gateways: map[v1alpha1.IngressVisibility]*GatewayConfig{
+				v1alpha1.IngressVisibilityExternalIP:   {GatewayClass: defaultGatewayClass, Gateway: defaultIstioGateway, Service: defaultGatewayService},
+				v1alpha1.IngressVisibilityClusterLocal: {GatewayClass: defaultGatewayClass, Gateway: defaultIstioLocalGateway, Service: defaultLocalGatewayService},
+			},
+		}, nil
+	}
+
+	entry := make(map[v1alpha1.IngressVisibility]GatewayConfig)
+	if err := yaml.Unmarshal([]byte(v), &entry); err != nil {
+		return nil, err
+	}
+
+	for _, vis := range []v1alpha1.IngressVisibility{
+		v1alpha1.IngressVisibilityClusterLocal,
+		v1alpha1.IngressVisibilityExternalIP,
+	} {
+		if _, ok := entry[vis]; !ok {
+			return nil, fmt.Errorf("visibility %q must not be empty", vis)
+		}
+	}
+	c := Gateway{Gateways: map[v1alpha1.IngressVisibility]*GatewayConfig{}}
+
+	for key, value := range entry {
+		key, value := key, value
+		// Check that the visibility makes sense.
+		switch key {
+		case v1alpha1.IngressVisibilityClusterLocal, v1alpha1.IngressVisibilityExternalIP:
+		default:
+			return nil, fmt.Errorf("unrecognized visibility: %q", key)
+		}
+
+		// See if the Service is a valid namespace/name token.
+		if _, _, err := cache.SplitMetaNamespaceKey(value.Service); err != nil {
+			return nil, err
+		}
+
+		// See if the Gateway is a valid namespace/name token.
+		if _, _, err := cache.SplitMetaNamespaceKey(value.Gateway); err != nil {
+			return nil, err
+		}
```

Comment:
From the comment at the using side: Should we store `types.NamespacedName` in `Gateways` type so that calling sides can use this without having to ignore errors and in a structured way already?